server/tests/audit_math/test_supersimple_raire.py

```diff
@@ -0,0 +1,1146 @@
+        sample_cvr[0] = {
+            "times_sampled": 1,
+            "cvr": {
+                "Contest A": {"winner": 0, "loser": 1},
+                "Contest B": {"winner": 0, "loser": 1},
+                "Contest C": {"winner": 0, "loser": 1},
+                "Contest D": {"winner": 0, "loser": 1},
+                "Contest E": {"winner": 0, "loser": 1},
+            },
+        }
+
+        p_value, finished = supersimple_raire.compute_risk(
+            RISK_LIMIT, contests[contest], cvrs, sample_cvr, computed_assertions
+        )
+        expected_p = expected_p_values["two_vote_over"][contest]
+        diff = abs(p_value - expected_p)
+
+        assert (
+            diff < 0.001
+        ), "Incorrect p-value. Expected {}, got {} in contest {}".format(
+            expected_p, p_value, contest
+        )
+
+        if contest != "Contest F":
+            assert not finished, "Audit shouldn't have finished but did!"
+
+        to_sample = {
+            "sample_size": sample_size,
+            "1-under": 0,
+            "1-over": 0,
+            "2-under": 0,
+            "2-over": 1,
+        }
+
+        next_sample_size = supersimple_raire.get_sample_sizes(
+            RISK_LIMIT, contests[contest], cvrs, to_sample, assertions[contest]
+        )
+        assert (
+            next_sample_size == o2_stopping_size[contest] - sample_size
+        ), "Number of ballots left to sample is not correct in contest {}!".format(
+            contest
+        )
+
+
+def parse_shangrla_sample(input_file: str) -> SAMPLECVRS:
+    cvrs = {}
+    with open(input_file, "r") as datafile:
+        sample_data = json.load(datafile)
+        ballots = sample_data["ballots"]
+
+        for ballot in ballots:
+            if ballot["id"] not in cvrs:
+                # For typechevker...
+                cvrs[ballot["id"]] = {
+                    "cvr": ballot["votes"],
+                    "times_sampled": 1,
+                }
+            else:
+                cvrs[ballot["id"]]["times_sampled"] += 1
+
+    return cast(SAMPLECVRS, cvrs)
+
+
+def test_get_sample_sizes_fptp(contests, cvrs, assertions):
+    for contest in contests:
+        computed = supersimple_raire.get_sample_sizes(
+            RISK_LIMIT, contests[contest], cvrs, None, assertions[contest]
+        )
+        expected = true_sample_sizes[contest]  # From Stark's tool
+
+        assert (
+            computed == expected
+        ), "Sample size computation incorrect: got {}, expected {} in contest {}".format(
+            computed, expected, contest
+        )
+
+
+def test_validate_cvr():
+
+    expected: CVR = {"contest": {"Alice": 1, "Bob": 2, "Charlie": 0}}
+
+    # Idempotent for correct cvrs
+    assert expected == supersimple_raire.validate_cvr(
+        {"contest": {"Alice": 1, "Bob": 2, "Charlie": 0}}
+    )
+
+    # Correctly fixes gaps
+    assert expected == supersimple_raire.validate_cvr(
+        {"contest": {"Alice": 2, "Bob": 4, "Charlie": 0}}
+    )
```

Comment:
I think this function could benefit from some more tests, since it's fairly complex logic. Currently there's no testing for the overvote adjustment logic, and only one case covered for the gap-fixing logic.

Reply:
Though I did just notice the next test function below which is awesome! Love that you tested from the discrepancy perspective

Reply:
I realized it was easier to test the discrepancies than doing it here, so that's what I went with. I can beef this one up too.